### DISCUSSION TITLE
Is it okay that a lot of tests are failing on the latest `main`?

### DISCUSSION THREAD

**Discussion by ansavchenco** [adobe/react-spectrum#3040] [Q&A] [answered]
I'm considering contributing to the this project and so I forked the repo and ran the tests with `yarn jest` command as described in the `CONTRIBUTING.md` and a lot of them failed. Here is the output.
```
Snapshot Summary
 › 1 snapshot failed from 1 test suite. Inspect your code changes or run `yarn run jest -u` to update them.

Test Suites: 23 failed, 2 skipped, 140 passed, 163 of 165 total
Tests:       532 failed, 55 skipped, 3318 passed, 3905 total
Snapshots:   1 failed, 1 total
Time:        112.172 s
```

Is this okay? How would I know that my code does not break anything? Should I only run the tests that are related to the part of the codebase I have made changes to? Or am I just doing something wrong?

Thanks!

**Comment by LFDanLu** [comment]
Odd, mind sharing maybe a couple of the failures your are seeing? The tests should be passing on main. What version of node are you running with?

**Reply by LFDanLu** [reply]
Another weird thing is the snapshot test failures, we don't have any snapshot tests in our codebase, did you add snapshot tests locally? Mind sharing what path you are executing `yarn jest` from (can use `pwd` to get this).

**Reply by ansavchenco** [reply]
Sure! My node version is `v14.17.6` and `yarn` is `1.22.10`. 
The path is `/Users/antonsavcenko/Development/oss/react-spectrum`

```
  ● SSRProvider › it should generate consistent unique ids with nested SSR providers

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `SSRProvider it should generate consistent unique ids with nested SSR providers 1`

    - Snapshot  - 3
    + Received  + 3

      Array [
        "react-aria-1",
    +   "react-aria-1-1",
    +   "react-aria-1-1-1",
    +   "react-aria-2",
        "react-aria-2-1",
    -   "react-aria-2-2-1",
    -   "react-aria-3",
    -   "react-aria-4-1",
      ]

      52 |
      53 |     let divs = tree.getAllByTestId('test');
    > 54 |     expect(divs.map((div) => div.id)).toMatchInlineSnapshot(`
         |                                       ^
      55 |       Array [
      56 |         "react-aria-1",
      57 |         "react-aria-2-1",

      at Object.<anonymous> (packages/@react-aria/ssr/test/SSRProvider.test.js:54:39)
```

```
 FAIL   component and hooks  packages/@react-spectrum/datepicker/test/DateField.test.js
  ● DateField › labeling › should support help text with a value

    expect(received).toBe(expected) // Object.is equality

    Expected: "Selected Date: February 3, 2020 Help text"
    Received: "February 3, 2020 Help text"

      134 |
      135 |       let description = group.getAttribute('aria-describedby').split(' ').map(d => document.getElementById(d).textContent).join(' ');
    > 136 |       expect(description).toBe('Selected Date: February 3, 2020 Help text');
          |                           ^
      137 |
      138 |       let segments = getAllByRole('spinbutton');
      139 |       expect(segments[0]).toHaveAttribute('aria-describedby', group.getAttribute('aria-describedby'));

      at Object.<anonymous> (packages/@react-spectrum/datepicker/test/DateField.test.js:136:27)
```

Also, a lot of tests are failing at something like `act(() => {jest.runAllTimers();});`.
```
  ● TableView › headerless columns › renders table with headerless column with tooltip

    Should not already be working.

      125 |
      126 |   afterEach(() => {
    > 127 |     act(() => {jest.runAllTimers();});
          |                     ^
      128 |   });
      129 |
      130 |   let render = (children, scale = 'medium') => renderComponent(

      at performSyncWorkOnRoot (node_modules/react-dom/cjs/react-dom.development.js:22265:13)
      at node_modules/react-dom/cjs/react-dom.development.js:11327:26
      at unstable_runWithPriority (node_modules/react-dom/node_modules/scheduler/cjs/scheduler.development.js:468:12)
      at runWithPriority$1 (node_modules/react-dom/cjs/react-dom.development.js:11276:10)
      at flushSyncCallbackQueueImpl (node_modules/react-dom/cjs/react-dom.development.js:11322:9)
      at flushSyncCallbackQueue (node_modules/react-dom/cjs/react-dom.development.js:11309:3)
      at workLoop (node_modules/react-dom/node_modules/scheduler/cjs/scheduler.development.js:417:34)
      at flushWork (node_modules/react-dom/node_modules/scheduler/cjs/scheduler.development.js:390:14)
      at _flushCallback (node_modules/react-dom/node_modules/scheduler/cjs/scheduler.development.js:55:9)
      at Object.callback (node_modules/@jest/fake-timers/build/legacyFakeTimers.js:625:32)
      at FakeTimers._runTimerHandle (node_modules/@jest/fake-timers/build/legacyFakeTimers.js:659:15)
      at FakeTimers.runAllTimers (node_modules/@jest/fake-timers/build/legacyFakeTimers.js:207:12)
      at packages/@react-spectrum/table/test/Table.test.js:127:21
      at batchedUpdates$1 (node_modules/react-dom/cjs/react-dom.development.js:22380:12)
      at act (node_modules/react-dom/cjs/react-dom-test-utils.development.js:1042:14)
      at Object.<anonymous> (packages/@react-spectrum/table/test/Table.test.js:127:5)
```

**Reply by ansavchenco** [reply]
I haven't changes anything locally. I also tried clearing yarn cache and reinstalling dependencies. The result is still the same.

**Reply by snowystinger** [reply]
What machine do you use? is it an M1 Mac? if so, I use node 16.13.2 currently, you could try that

**Reply by snowystinger** [reply]
Hmmm I tried running it too, everything passes for me in a brand new clone of the repo.
I know you said you haven't changed anything.

Could you triple check that you have no local changes in your fork, committed or otherwise? Or even better, clone the repo anew, and run the tests from the new clone?

Apologies if you've been through this, I just really want to make sure.

**Reply by LFDanLu** [reply]
Just an update, @snowystinger and I just finished running the tests on a variety of setups (on `main` and a fork of the repro with a variety of node versions) and they all pass. The tests are passing just fine on [CircleCi as well](https://app.circleci.com/pipelines/github/adobe/react-spectrum/8364/workflows/056732b7-de90-4a57-9327-0d46436020eb/jobs/78494), so we believe there might be something different in your environment. I'm not entirely sure what could be causing these issues, could you also try running the failing tests one by one in case this might be a cascading test failure issue?

**Reply by ansavchenco** [reply]
I'm on Intel Mac and recloning the repo helped. On node `v14.17.6` I still have 2 failing tests but on `v16.11.0` it's all green!

Here are the tests that failed on `v14.17.6`.
```
 FAIL  packages/@react-spectrum/numberfield/test/NumberField.test.js (56.794 s)
  ● NumberField › nb-NO can have arabic numerals entered

    expect(element).toHaveAttribute("value", "US$21.00") // element.getAttribute("value") === "US$21.00"

    Expected the element to have attribute:
      value="US$21.00"
    Received:
      value=""

      1920 |
      1921 |     let formatter = new Intl.NumberFormat(locale + '-u-nu-arab', {style: 'currency', currency: 'USD'});
    > 1922 |     expect(textField).toHaveAttribute('value', formatter.format(21));
           |                       ^
      1923 |   });
      1924 |
      1925 |   it.each(locales)('%s can have hanidec numerals entered', (locale) => {

      at packages/@react-spectrum/numberfield/test/NumberField.test.js:1922:23

  ● NumberField › nb-NO can have hanidec numerals entered

    expect(element).toHaveAttribute("value", "US$21.00") // element.getAttribute("value") === "US$21.00"

    Expected the element to have attribute:
      value="US$21.00"
    Received:
      value=""

      1931 |
      1932 |     let formatter = new Intl.NumberFormat(locale + '-u-nu-hanidec', {style: 'currency', currency: 'USD'});
    > 1933 |     expect(textField).toHaveAttribute('value', formatter.format(21));
           |                       ^
      1934 |   });
      1935 |
      1936 |   describe('beforeinput', () => {

      at packages/@react-spectrum/numberfield/test/NumberField.test.js:1933:23
```

Probably the problem was that I actually did make changes to code and then reverted them before running the tests.
Anyway, thank you all for your help!

**Reply by snowystinger** [reply]
Glad it's working now. The two you're seeing fail are known, some node 14 versions have a bug in Intl unfortunately and it's not worth fixing because they work everywhere else.

**Comment by ansavchenco** [comment, marked as answer]
As @[snowystinger](https://github.com/snowystinger) suggested [here](https://github.com/adobe/react-spectrum/discussions/3040#discussioncomment-2570133) I recloned the repo and switched to node@v16 and it solved the problem.

**Comment by piecyk** [comment]
Hmm same for me, lot's of failing tests, was trying node v16.14.0, v18.15.0 yarn 1.22.19 on M1 Mac

running for example `yarn test -- react-aria-components/test/TagGroup.test.js` 

>  Error: Uncaught [Error: Invalid slot "description". Valid slot names are "description" i "errorMessage".]

build also is not working 🤷‍♂️ 

```
🚨 Build failed.

unknown: Could not find entry: /Users/dpieczynski/work/my/react-spectrum/packages/@react-spectrum/alert
make: *** [build] Error 1
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Reply by LFDanLu** [reply]
Does node 18.17.1 work for you?

**Reply by piecyk** [reply]
No, really strange. 

```
➜  react-spectrum git:(main) ✗ node -v && yarn -v
v18.17.1
1.22.19
➜  react-spectrum git:(main) ✗ yarn build
yarn run v1.22.19
$ make build
parcel build packages/@react-{spectrum,aria,stately}/*/ packages/@internationalized/{message,string,date,number}/ packages/react-aria-components --no-optimize
🚨 Build failed.

unknown: Could not find entry: /Users/dpieczynski/work/my/react-spectrum/packages/@react-spectrum/alert
make: *** [build] Error 1
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
➜  react-spectrum git:(main) ✗
```

**Reply by LFDanLu** [reply]
Definitely odd, worked for me on a fresh clone of the repo. @snowystinger did you need to do anything special for you M1 Mac?

**Reply by snowystinger** [reply]
Nope, I use zsh and just pulled a fresh clone as well. Both commands worked.
Could you try switching to node 18, pulling a new copy of the repo down, and running the commands again. I just want to rule out anything being left behind in the switch from node 16 -> 18

**Reply by piecyk** [reply]
Ok, also tried fresh clone, latest node 18, setup with nvm, 

* build worked, there where some type errors 
* after running,  yarn run test

```
Test Suites: 24 failed, 188 passed, 212 total
Tests:       1444 failed, 7 skipped, 3877 passed, 5328 total
Snapshots:   0 total
Time:        46.541 s
```

still feels not right

**Reply by snowystinger** [reply]
I think it has to do with order of operations. You're running build, which is creating dist folders in all of the packages. When node does resolutions during testing, it's finding those dist folders which aren't what the tests expect.
If you remove all the dist folders, it should work. You can do a `make clean_dist` to clear them out.

**Reply by snowystinger** [reply]
just checking in, were you able to get it working?
